### DISCUSSION TITLE
Teensy 4, problem with encoders under Mcp2301

### DISCUSSION THREAD

**Discussion by antomexita** [tttapa/Control-Surface#751] [Q&A] [answered]
Hi to all,
My teensy 4 is connect with Mcp23017 that drive 8 encoders. The encoders responding in the serial monitor but the value is only 1  or 127 like a push button.  How to make a incremental / decremental reading ?
In attach the code
Gracias 

`#include <Control_Surface.h>
#include <AH/Hardware/MCP23017Encoders.hpp>
#include <Wire.h>
using WireType = decltype(Wire);
using EncoderPositionType = int32_t;
constexpr bool IntSafe = true;
using MCPEncoderType = MCP23017Encoders<WireType, EncoderPositionType, IntSafe>;
struct CCMCPEncoder : GenericMIDIRotaryEncoder<MCPEncoderType::MCP23017Encoder,
  RelativeCCSender> {
  CCMCPEncoder(MCPEncoderType::MCP23017Encoder enc, MIDIAddress address,
               int16_t multiplier = 1, uint8_t pulsesPerStep = 4)
    : GenericMIDIRotaryEncoder(std::move(enc), address, multiplier,
                               pulsesPerStep, {}) {}
};
USBMIDI_Interface midi;
const uint8_t interrupt_pin = 9;
MCPEncoderType enc {Wire, 0x20, interrupt_pin};
CCMCPEncoder ccencoders[] {
  {enc[0], {44, CHANNEL_16}, 2, 4},
  {enc[1], MCU::V_POT_2},
  {enc[2], MCU::V_POT_3},
  {enc[3], MCU::V_POT_4},
  {enc[4], MCU::V_POT_5},
  {enc[5], MCU::V_POT_6},
  {enc[6], MCU::V_POT_7},
  {enc[7], MCU::V_POT_8},
};
SPIShiftRegisterOut<8> sreg {
  SPI,
  10,
  MSBFIRST,
};
NoteLEDBar<8> leds { sreg.pins(), MIDI_Notes::C(4) };
void isr() {
  enc.update();
}
void setup() {
  Control_Surface.begin();
  Wire.begin();
  Wire.setClock(800000);
  enc.begin();
  RelativeCCSender::setMode(TWOS_COMPLEMENT);
  attachInterrupt(digitalPinToInterrupt(interrupt_pin), isr, LOW);
}

void loop() {
  enc.update();
  Control_Surface.loop();
}`

**Comment by Veroledez** [comment]
Hola, prueba con la clase absoluteencoder para obtener valores de 0 a 127 como si fuese un potenciometro

**Reply by antomexita** [reply]
intente cambiar GenericMIDIEncoder con GenericMIDIAbsoluteEncoder pero no funciona.

**Reply by tttapa** [reply]
> pero no funciona.

Please post the exact code you used and the full error message you get.

Did you use the correct MIDI sender? E.g. `GenericMIDIAbsoluteEncoder<MCPEncoderType::MCP23017Encoder, ContinuousCCSender>`?

**Comment by tttapa** [comment, marked as answer]
> the value is only 1 or 127 like a push button. How to make a incremental / decremental reading ?

This _is_ the incremental reading: 1 is a positive increment and 127 is a negative increment.  
See also [relativeCCmode](https://tttapa.github.io/Control-Surface-doc/Doxygen/d9/d24/RelativeCCSender_8hpp.html#a793b4d05f50a2c3af3ddeeef50e7a71d).

If you want an absolute reading instead of an incremental reading, have a look at the `AbsoluteEncoder` classes, e.g. [`GenericMIDIAbsoluteEncoder`](https://tttapa.github.io/Control-Surface-doc/Doxygen/df/d5d/classGenericMIDIAbsoluteEncoder.html).

**Reply by antomexita** [reply]
Hi Pieter, 
"ContinuousCCSender" resolve it!
a ContinuousManyThanks  for the rest of the day.  

Here the code whitout interrupt funcion.   

`#include <Control_Surface.h>
#include <AH/Hardware/MCP23017Encoders.hpp>
#include <Wire.h>
using WireType = decltype(Wire);
using EncoderPositionType = uint8_t;
using MCPEncoderType = MCP23017Encoders<WireType, EncoderPositionType>;
struct CCMCPEncoder : GenericMIDIAbsoluteEncoder<MCPEncoderType::MCP23017Encoder,
                                               ContinuousCCSender> {
  CCMCPEncoder(MCPEncoderType::MCP23017Encoder enc, MIDIAddress address,
               int16_t multiplier = 1, uint8_t pulsesPerStep = 4)
    : GenericMIDIAbsoluteEncoder(std::move(enc), address, multiplier,
                               pulsesPerStep, {}) {}
};
USBMIDI_Interface midi;
MCPEncoderType enc {Wire, 0x20, 9};
CCMCPEncoder ccencoders[] {
  {enc[0], {MIDI_CC::Modulation_Wheel, CHANNEL_1}, 2, 4},
  {enc[1], {MIDI_CC::Breath_Controller, CHANNEL_1}, 2, 4},
  {enc[2], {MIDI_CC::Foot_Controller, CHANNEL_1}, 2, 4},
  {enc[3], {MIDI_CC::Portamento_Time, CHANNEL_1}, 2, 4},
  {enc[4], {MIDI_CC::Balance, CHANNEL_1}, 2, 4},
  {enc[5], {MIDI_CC::Pan, CHANNEL_1}, 2, 4},
  {enc[6], {MIDI_CC::Expression_Controller, CHANNEL_1}, 2, 4},
  {enc[7], {MIDI_CC::Portamento, CHANNEL_1}, 2, 4},
};
SPIShiftRegisterOut<8> sreg {
  SPI,
  10,
  MSBFIRST,
};
NoteLEDBar<8> leds { sreg.pins(), MIDI_Notes::C(4) };
void setup() {
  Control_Surface.begin();
  Wire.begin();
  Wire.setClock(800000);
  enc.begin();
}

void loop() {
  enc.update();
  Control_Surface.loop();
}`

**Comment by antomexita** [comment]
Hi Pieter,
i have a little problem with this encoders. I make the pcb but   when  rotate, some encoders instead of increasing the value, they decrease and reverse. They are connected in reverse. To avoid making a new pcb where can I change the order of the pins in order to make them work correctly? I hope it is doable.
Thanks in advance

**Reply by tttapa** [reply]
You can simply set the `multiplier` argument to -1 instead of 1.
